File: artifacts/api-server/src/routes/notifications.ts
import { Router, type IRouter } from "express";
import { TestTelegramResponse } from "@workspace/api-zod";
import { sendTestMessage } from "../services/telegram";
import { getCachedMarkets } from "../services/polymarket";
import { FAKE_MARKETS } from "./markets";
import db from "../lib/db";

const router: IRouter = Router();

interface AlertRow {
  id: number;
  market_id: string;
  market_question: string;
  side: string;
  direction: string;
  target_price: number;
  triggered: number;
  triggered_at: string | null;
  created_at: string;
}

interface WatchlistRow {
  market_id: string;
  market_question: string;
  category: string;
  yes_price: number;
  no_price: number;
  volume24h: number;
  added_at: string;
}

router.post("/telegram/test", async (_req, res) => {
  const result = await sendTestMessage();
  res.json(TestTelegramResponse.parse(result));
});

router.get("/alerts", (req, res) => {
  const includeTriggered = req.query.triggered === "true";
  const rows = db.prepare(
    `SELECT id, market_id, market_question, side, direction, target_price, triggered, triggered_at, created_at
     FROM price_alerts WHERE triggered = ${includeTriggered ? "1 OR triggered = 0" : "0"}
     ORDER BY created_at DESC`
  ).all() as AlertRow[];
  res.json(rows.map((r) => ({
    id: r.id, marketId: r.market_id, marketQuestion: r.market_question, side: r.side,
    direction: r.direction, targetPrice: r.target_price, triggered: r.triggered === 1,
    triggeredAt: r.triggered_at, createdAt: r.created_at,
  })));
});

router.post("/alerts", (req, res) => {
  const { marketId, marketQuestion, side, direction, targetPrice } = req.body as { marketId?: string; marketQuestion?: string; side?: string; direction?: string; targetPrice?: number };
  if (!marketId || !side || !direction || targetPrice === undefined) {
    res.status(400).json({ error: "marketId, side, direction, targetPrice required" }); return;
  }
  if (!["YES", "NO"].includes(side)) { res.status(400).json({ error: "side must be YES or NO" }); return; }
  if (!["above", "below"].includes(direction)) { res.status(400).json({ error: "direction must be above or below" }); return; }
  if (typeof targetPrice !== "number" || targetPrice <= 0 || targetPrice >= 1) {
    res.status(400).json({ error: "targetPrice must be between 0 and 1" }); return;
  }
  const result = db.prepare(
    "INSERT INTO price_alerts (market_id, market_question, side, direction, target_price, triggered, created_at) VALUES (?, ?, ?, ?, ?, 0, ?)"
  ).run(marketId, marketQuestion ?? `Market ${marketId}`, side, direction, targetPrice, new Date().toISOString());
  res.json({ success: true, id: result.lastInsertRowid });
});

router.delete("/alerts/:id", (req, res) => {
  const id = parseInt(req.params.id, 10);
  if (isNaN(id)) { res.status(400).json({ error: "Invalid alert ID" }); return; }
  const result = db.prepare("DELETE FROM price_alerts WHERE id = ?").run(id);
  if (result.changes === 0) { res.status(404).json({ error: "Alert not found" }); return; }
  res.json({ success: true });
});

router.get("/watchlist", (_req, res) => {
  const rows = db.prepare(
    "SELECT market_id, market_question, category, yes_price, no_price, volume24h, added_at FROM market_watchlist ORDER BY added_at DESC"
  ).all() as WatchlistRow[];
  res.json(rows.map((r) => ({
    marketId: r.market_id, marketQuestion: r.market_question, category: r.category,
    yesPrice: r.yes_price, noPrice: r.no_price, volume24h: r.volume24h, addedAt: r.added_at,
  })));
});

router.post("/watchlist", async (req, res) => {
  const { marketId } = req.body as { marketId?: string };
  if (!marketId || typeof marketId !== "string") { res.status(400).json({ error: "marketId required" }); return; }
  let question = "", category = "", yesPrice = 0.5, noPrice = 0.5, volume24h = 0;
  try {
    const live = await getCachedMarkets();
    const found = live.find((m) => m.id === marketId);
    if (found) { question = found.question; category = found.category; yesPrice = found.yesPrice; noPrice = found.noPrice; volume24h = found.volume24h; }
  } catch { }
  if (!question) {
    const demo = FAKE_MARKETS.find((m) => m.id === marketId);
    if (demo) { question = demo.question; category = demo.category; yesPrice = demo.yesPrice; noPrice = demo.noPrice; volume24h = demo.volume24h; }
  }
  if (!question) question = `Market ${marketId}`;
  db.prepare(
    "INSERT OR REPLACE INTO market_watchlist (market_id, market_question, category, yes_price, no_price, volume24h, added_at) VALUES (?, ?, ?, ?, ?, ?, ?)"
  ).run(marketId, question, category, yesPrice, noPrice, volume24h, new Date().toISOString());
  res.json({ success: true, marketId, marketQuestion: question });
});

router.delete("/watchlist/:marketId", (req, res) => {
  const { marketId } = req.params;
  const result = db.prepare("DELETE FROM market_watchlist WHERE market_id = ?").run(marketId);
  if (result.changes === 0) { res.status(404).json({ error: "Market not in watchlist" }); return; }
  res.json({ success: true, marketId });
});

router.get("/watchlist/:marketId", (req, res) => {
  const { marketId } = req.params;
  const row = db.prepare("SELECT market_id FROM market_watchlist WHERE market_id = ?").get(marketId) as { market_id: string } | undefined;
  res.json({ watched: !!row });
});

export default router;

File: server/internal/routes/notifications.go
package routes

import (
	"database/sql"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"polymarket-trader/internal/db"
	"polymarket-trader/internal/services"
)

func RegisterNotificationsRoutes(rg *gin.RouterGroup) {
	g := rg.Group("/notifications")
	g.GET("/status", getNotificationStatus)
	g.POST("/test", sendTestNotification)
	g.GET("/watchlist", getWatchlist)
	g.POST("/watchlist", addToWatchlist)
	g.DELETE("/watchlist/:id", removeFromWatchlist)
	g.GET("/alerts", getPriceAlerts)
	g.POST("/alerts", createPriceAlert)
	g.DELETE("/alerts/:id", deletePriceAlert)
}

func getNotificationStatus(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"telegramConfigured": services.IsTelegramConfigured(),
		"hasToken":           services.BotToken() != "",
		"hasChatID":          services.TelegramChatID() != "",
	})
}

func sendTestNotification(c *gin.Context) {
	result := services.SendTestMessage()
	status := http.StatusOK
	if !result["success"].(bool) {
		status = http.StatusBadRequest
	}
	c.JSON(status, result)
}

func getWatchlist(c *gin.Context) {
	rows, err := db.DB.Query("SELECT market_id, market_question, category, yes_price, no_price, volume24h, added_at FROM market_watchlist ORDER BY added_at DESC")
	if err != nil {
		c.JSON(http.StatusOK, gin.H{"watchlist": []interface{}{}})
		return
	}
	defer rows.Close()
	var items []map[string]interface{}
	for rows.Next() {
		var marketID, question, category, addedAt string
		var yesPrice, noPrice, volume24h float64
		rows.Scan(&marketID, &question, &category, &yesPrice, &noPrice, &volume24h, &addedAt)
		items = append(items, map[string]interface{}{
			"marketId":       marketID,
			"marketQuestion": question,
			"category":       category,
			"yesPrice":       yesPrice,
			"noPrice":        noPrice,
			"volume24h":      volume24h,
			"addedAt":        addedAt,
		})
	}
	if items == nil {
		items = []map[string]interface{}{}
	}
	c.JSON(http.StatusOK, gin.H{"watchlist": items})
}

func addToWatchlist(c *gin.Context) {
	var body struct {
		MarketID string `json:"marketId"`
		Question string `json:"marketQuestion"`
		Category string `json:"category"`
		YesPrice float64 `json:"yesPrice"`
		NoPrice  float64 `json:"noPrice"`
		Volume24h float64 `json:"volume24h"`
	}
	if err := c.ShouldBindJSON(&body); err != nil || body.MarketID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "marketId is required"})
		return
	}
	now := time.Now().Format(time.RFC3339)
	_, err := db.DB.Exec(
		"INSERT OR REPLACE INTO market_watchlist (market_id, market_question, category, yes_price, no_price, volume24h, added_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
		body.MarketID, body.Question, body.Category, body.YesPrice, body.NoPrice, body.Volume24h, now,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add to watchlist"})
		return
	}
	c.JSON(http.StatusCreated, gin.H{"message": "Added to watchlist", "marketId": body.MarketID})
}

func removeFromWatchlist(c *gin.Context) {
	id := c.Param("id")
	res, err := db.DB.Exec("DELETE FROM market_watchlist WHERE market_id = ?", id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to remove"})
		return
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Market not in watchlist"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Removed from watchlist"})
}

func getPriceAlerts(c *gin.Context) {
	rows, err := db.DB.Query("SELECT id, market_id, market_question, side, direction, target_price, triggered, triggered_at, created_at FROM price_alerts ORDER BY created_at DESC")
	if err != nil {
		c.JSON(http.StatusOK, gin.H{"alerts": []interface{}{}})
		return
	}
	defer rows.Close()
	var items []map[string]interface{}
	for rows.Next() {
		var id int64
		var marketID, question, side, direction, createdAt string
		var targetPrice float64
		var triggered int
		var triggeredAt sql.NullString
		rows.Scan(&id, &marketID, &question, &side, &direction, &targetPrice, &triggered, &triggeredAt, &createdAt)
		item := map[string]interface{}{
			"id":             id,
			"marketId":       marketID,
			"marketQuestion": question,
			"side":           side,
			"direction":      direction,
			"targetPrice":    targetPrice,
			"triggered":      triggered == 1,
			"createdAt":      createdAt,
		}
		if triggeredAt.Valid {
			item["triggeredAt"] = triggeredAt.String
		}
		items = append(items, item)
	}
	if items == nil {
		items = []map[string]interface{}{}
	}
	c.JSON(http.StatusOK, gin.H{"alerts": items})
}

func createPriceAlert(c *gin.Context) {
	var body struct {
		MarketID  string  `json:"marketId"`
		Question  string  `json:"marketQuestion"`
		Side      string  `json:"side"`
		Direction string  `json:"direction"`
		TargetPrice float64 `json:"targetPrice"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid body"})
		return
	}
	if body.MarketID == "" || body.Side == "" || body.Direction == "" || body.TargetPrice <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "marketId, side, direction, and targetPrice are required"})
		return
	}
	now := time.Now().Format(time.RFC3339)
	res, err := db.DB.Exec(
		"INSERT INTO price_alerts (market_id, market_question, side, direction, target_price, triggered, created_at) VALUES (?, ?, ?, ?, ?, 0, ?)",
		body.MarketID, body.Question, body.Side, body.Direction, body.TargetPrice, now,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create alert"})
		return
	}
	id, _ := res.LastInsertId()
	c.JSON(http.StatusCreated, gin.H{"id": id, "message": "Price alert created"})
}

func deletePriceAlert(c *gin.Context) {
	id := c.Param("id")
	res, err := db.DB.Exec("DELETE FROM price_alerts WHERE id = ?", id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete"})
		return
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Alert not found"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Alert deleted"})
}
